stty/stty.1
.Dd December 2008
.Dt STTY 1

.Sh NAME

.Nm stty
.Nd set the options for a terminal

.Sh SYNOPSIS

.Nm stty
.Ar [ -a| -g]

.Nm stty
.Ar operands

.Sh DESCRIPTION

    The stty utility shall set or report on terminal I/O characteristics for
the device that is its standard input. Without options or operands specified,
it shall report the settings of certain characteristics, usually those that
differ from implementation-defined defaults. Otherwise, it shall modify the
terminal state according to the specified operands. Detailed information
about the modes listed in the first five groups below are described in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface. Operands in the Combination Modes group (see Combination Modes)
are implemented using operands in the previous groups. Some combinations of
operands are mutually-exclusive on some terminal types; the results of using
such combinations are unspecified.

    Typical implementations of this utility require a communications line
configured to use the termios interface defined in the System Interfaces
volume of IEEE Std 1003.1-2001. On systems where none of these lines are
available, and on lines not currently configured to support the termios
interface, some of the operands need not affect terminal characteristics.

.Sh OPTIONS

    The stty utility shall conform to the Base Definitions volume of IEEE Std
1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -a
        Write to standard output all the current settings for the terminal.
    -g
        Write to standard output all the current settings in an unspecified
form that can be used as arguments to another invocation of the stty utility
on the same system. The form used shall not contain any characters that would
require quoting to avoid word expansion by the shell; see Word Expansions.

.Sh OPERANDS

    The following operands shall be supported to set the terminal
characteristics.
    Control Modes

    parenb  (-parenb)
        Enable (disable) parity generation and detection. This shall have the
effect of setting (not setting) PARENB in the termios c_cflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    parodd  (-parodd)
        Select odd (even) parity. This shall have the effect of setting (not
setting) PARODD in the termios c_cflag field, as defined in the Base
Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    cs5 cs6 cs7 cs8
        Select character size, if possible. This shall have the effect of
setting CS5, CS6, CS7, and CS8, respectively, in the termios c_cflag field,
as defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter
11, General Terminal Interface.
    number
        Set terminal baud rate to the number given, if possible. If the baud
rate is set to zero, the modem control lines shall no longer be asserted.
This shall have the effect of setting the input and output termios baud rate
values as defined in the Base Definitions volume of IEEE Std 1003.1-2001,
Chapter 11, General Terminal Interface.
    ispeed  number
        Set terminal input baud rate to the number given, if possible. If the
input baud rate is set to zero, the input baud rate shall be specified by the
value of the output baud rate. This shall have the effect of setting the
input termios baud rate values as defined in the Base Definitions volume of
IEEE Std 1003.1-2001, Chapter 11, General Terminal Interface.
    ospeed  number
        Set terminal output baud rate to the number given, if possible. If
the output baud rate is set to zero, the modem control lines shall no longer
be asserted. This shall have the effect of setting the output termios baud
rate values as defined in the Base Definitions volume of IEEE Std
1003.1-2001, Chapter 11, General Terminal Interface.
    hupcl  (-hupcl)
        Stop asserting modem control lines (do not stop asserting modem
control lines) on last close. This shall have the effect of setting (not
setting) HUPCL in the termios c_cflag field, as defined in the Base
Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    hup  (-hup)
        Equivalent to hupcl( -hupcl).
    cstopb  (-cstopb)
        Use two (one) stop bits per character. This shall have the effect of
setting (not setting) CSTOPB in the termios c_cflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    cread  (-cread)
        Enable (disable) the receiver. This shall have the effect of setting
(not setting) CREAD in the termios c_cflag field, as defined in the Base
Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    clocal  (-clocal)
        Assume a line without (with) modem control. This shall have the
effect of setting (not setting) CLOCAL in the termios c_cflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.

    It is unspecified whether stty shall report an error if an attempt to set
a Control Mode fails.
    Input Modes

    ignbrk  (-ignbrk)
        Ignore (do not ignore) break on input. This shall have the effect of
setting (not setting) IGNBRK in the termios c_iflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    brkint  (-brkint)
        Signal (do not signal) INTR on break. This shall have the effect of
setting (not setting) BRKINT in the termios c_iflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    ignpar  (-ignpar)
        Ignore (do not ignore) bytes with parity errors. This shall have the
effect of setting (not setting) IGNPAR in the termios c_iflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    parmrk  (-parmrk)
        Mark (do not mark) parity errors. This shall have the effect of
setting (not setting) PARMRK in the termios c_iflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    inpck  (-inpck)
        Enable (disable) input parity checking. This shall have the effect of
setting (not setting) INPCK in the termios c_iflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    istrip  (-istrip)
        Strip (do not strip) input characters to seven bits. This shall have
the effect of setting (not setting) ISTRIP in the termios c_iflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    inlcr  (-inlcr)
        Map (do not map) NL to CR on input. This shall have the effect of
setting (not setting) INLCR in the termios c_iflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    igncr (-igncr)
        Ignore (do not ignore) CR on input. This shall have the effect of
setting (not setting) IGNCR in the termios c_iflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    icrnl  (-icrnl)
        Map (do not map) CR to NL on input. This shall have the effect of
setting (not setting) ICRNL in the termios c_iflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    ixon  (-ixon)
        Enable (disable) START/STOP output control. Output from the system is
stopped when the system receives STOP and started when the system receives
START. This shall have the effect of setting (not setting) IXON in the
termios c_iflag field, as defined in the Base Definitions volume of IEEE Std
1003.1-2001, Chapter 11, General Terminal Interface.
    ixany  (-ixany)
        [XSI] [Option Start] Allow any character to restart output. This
shall have the effect of setting (not setting) IXANY in the termios c_iflag
field, as defined in the Base Definitions volume of IEEE Std 1003.1-2001,
Chapter 11, General Terminal Interface. [Option End]
    ixoff  (-ixoff)
        Request that the system send (not send) STOP characters when the
input queue is nearly full and START characters to resume data transmission.
This shall have the effect of setting (not setting) IXOFF in the termios
c_iflag field, as defined in the Base Definitions volume of IEEE Std
1003.1-2001, Chapter 11, General Terminal Interface.

    Output Modes

    opost  (-opost)
        Post-process output (do not post-process output; ignore all other
output modes). This shall have the effect of setting (not setting) OPOST in
the termios c_oflag field, as defined in the Base Definitions volume of IEEE
Std 1003.1-2001, Chapter 11, General Terminal Interface.
    ocrnl  (-ocrnl)
        [XSI] [Option Start] Map (do not map) CR to NL on output This shall
have the effect of setting (not setting) OCRNL in the termios c_oflag field,
as defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter
11, General Terminal Interface.
    onocr  (-onocr)
        Do not (do) output CR at column zero. This shall have the effect of
setting (not setting) ONOCR in the termios c_oflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    onlret  (-onlret)
        The terminal newline key performs (does not perform) the CR function.
This shall have the effect of setting (not setting) ONLRET in the termios
c_oflag field, as defined in the Base Definitions volume of IEEE Std
1003.1-2001, Chapter 11, General Terminal Interface.
    ofill  (-ofill)
        Use fill characters (use timing) for delays. This shall have the
effect of setting (not setting) OFILL in the termios c_oflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    ofdel  (-ofdel)
        Fill characters are DELs (NULs). This shall have the effect of
setting (not setting) OFDEL in the termios c_oflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    cr0 cr1 cr2 cr3
        Select the style of delay for CRs. This shall have the effect of
setting CRDLY to CR0, CR1, CR2, or CR3, respectively, in the termios c_oflag
field, as defined in the Base Definitions volume of IEEE Std 1003.1-2001,
Chapter 11, General Terminal Interface.
    nl0 nl1
        Select the style of delay for NL. This shall have the effect of
setting NLDLY to NL0 or NL1, respectively, in the termios c_oflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    tab0 tab1 tab2 tab3
        Select the style of delay for horizontal tabs. This shall have the
effect of setting TABDLY to TAB0, TAB1, TAB2, or TAB3, respectively, in the
termios c_oflag field, as defined in the Base Definitions volume of IEEE Std
1003.1-2001, Chapter 11, General Terminal Interface. Note that TAB3 has the
effect of expanding <tab>s to <space>s.
    tabs  (-tabs)
        Synonym for tab0 ( tab3).
    bs0 bs1
        Select the style of delay for backspaces. This shall have the effect
of setting BSDLY to BS0 or BS1, respectively, in the termios c_oflag field,
as defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter
11, General Terminal Interface.
    ff0 ff1
        Select the style of delay for form-feeds. This shall have the effect
of setting FFDLY to FF0 or FF1, respectively, in the termios c_oflag field,
as defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter
11, General Terminal Interface.
    vt0 vt1
        Select the style of delay for vertical-tabs. This shall have the
effect of setting VTDLY to VT0 or VT1, respectively, in the termios c_oflag
field, as defined in the Base Definitions volume of IEEE Std 1003.1-2001,
Chapter 11, General Terminal Interface. [Option End]

    Local Modes

    isig  (-isig)
        Enable (disable) the checking of characters against the special
control characters INTR, QUIT, and SUSP. This shall have the effect of
setting (not setting) ISIG in the termios c_lflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.
    icanon  (-icanon)
        Enable (disable) canonical input (ERASE and KILL processing). This
shall have the effect of setting (not setting) ICANON in the termios c_lflag
field, as defined in the Base Definitions volume of IEEE Std 1003.1-2001,
Chapter 11, General Terminal Interface.
    iexten  (-iexten)
        Enable (disable) any implementation-defined special control
characters not currently controlled by icanon, isig, ixon, or ixoff. This
shall have the effect of setting (not setting) IEXTEN in the termios c_lflag
field, as defined in the Base Definitions volume of IEEE Std 1003.1-2001,
Chapter 11, General Terminal Interface.
    echo  (-echo)
        Echo back (do not echo back) every character typed. This shall have
the effect of setting (not setting) ECHO in the termios c_lflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    echoe  (-echoe)
        The ERASE character visually erases (does not erase) the last
character in the current line from the display, if possible. This shall have
the effect of setting (not setting) ECHOE in the termios c_lflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    echok  (-echok)
        Echo (do not echo) NL after KILL character. This shall have the
effect of setting (not setting) ECHOK in the termios c_lflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    echonl  (-echonl)
        Echo (do not echo) NL, even if echo is disabled. This shall have the
effect of setting (not setting) ECHONL in the termios c_lflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    noflsh  (-noflsh)
        Disable (enable) flush after INTR, QUIT, SUSP. This shall have the
effect of setting (not setting) NOFLSH in the termios c_lflag field, as
defined in the Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface.
    tostop  (-tostop)
        Send SIGTTOU for background output. This shall have the effect of
setting (not setting) TOSTOP in the termios c_lflag field, as defined in the
Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11, General Terminal
Interface.

    Special Control Character Assignments

    <control>-character string
        Set <control>-character to string. If <control>-character is one of
the character sequences in the first column of the following table, the
corresponding Base Definitions volume of IEEE Std 1003.1-2001, Chapter 11,
General Terminal Interface control character from the second column shall be
recognized. This has the effect of setting the corresponding element of the
termios c_cc array (see the Base Definitions volume of IEEE Std 1003.1-2001,
Chapter 13, Headers, <termios.h>).
        Table: Control Character Names in stty

        Control Character
        	

        c_cc Subscript
        	

        Description

        eof
        	

        VEOF
        	

        EOF character

        eol
        	

        VEOL
        	

        EOL character

        erase
        	

        VERASE
        	

        ERASE character

        intr
        	

        VINTR
        	

        INTR character

        kill
        	

        VKILL
        	

        KILL character

        quit
        	

        VQUIT
        	

        QUIT character

        susp
        	

        VSUSP
        	

        SUSP character

        start
        	

        VSTART
        	

        START character

        stop
        	

        VSTOP
        	

        STOP character

        If string is a single character, the control character shall be set
to that character. If string is the two-character sequence "^-" or the string
undef, the control character shall be set to _POSIX_VDISABLE , if it is in
effect for the device; if _POSIX_VDISABLE is not in effect for the device, it
shall be treated as an error. In the POSIX locale, if string is a
two-character sequence beginning with circumflex ( '^' ), and the second
character is one of those listed in the "^c" column of the following table,
the control character shall be set to the corresponding character value in
the Value column of the table.
        Table: Circumflex Control Characters in stty

        ^c
        	

        Value
        	

        ^c
        	

        Value
        	

        ^c
        	

        Value

        a, A
        	

        <SOH>
        	

        l, L
        	

        <FF>
        	

        w, W
        	

        <ETB>
        	

         
        	

         
        	

         
        	

         
        	

         

        b, B
        	

        <STX>
        	

        m, M
        	

        <CR>
        	

        x, X
        	

        <CAN>
        	

         
        	

         
        	

         
        	

         
        	

         

        c, C
        	

        <ETX>
        	

        n, N
        	

        <SO>
        	

        y, Y
        	

        <EM>
        	

         
        	

         
        	

         
        	

         
        	

         

        d, D
        	

        <EOT>
        	

        o, O
        	

        <SI>
        	

        z, Z
        	

        <SUB>
        	

         
        	

         
        	

         
        	

         
        	

         

        e, E
        	

        <ENQ>
        	

        p, P
        	

        <DLE>
        	

        [
        	

        <ESC>
        	

         
        	

         
        	

         
        	

         
        	

         

        f, F
        	

        <ACK>
        	

        q, Q
        	

        <DC1>
        	

        \
        	

        <FS>
        	

         
        	

         
        	

         
        	

         
        	

         

        g, G
        	

        <BEL>
        	

        r, R
        	

        <DC2>
        	

        ]
        	

        <GS>
        	

         
        	

         
        	

         
        	

         
        	

         

        h, H
        	

        <BS>
        	

        s, S
        	

        <DC3>
        	

        ^
        	

        <RS>
        	

         
        	

         
        	

         
        	

         
        	

         

        i, I
        	

        <HT>
        	

        t, T
        	

        <DC4>
        	

        _
        	

        <US>
        	

         
        	

         
        	

         
        	

         
        	

         

        j, J
        	

        <LF>
        	

        u, U
        	

        <NAK>
        	

        ?
        	

        <DEL>
        	

         
        	

         
        	

         
        	

         
        	

         

        k, K
        	

        <VT>
        	

        v, V
        	

        <SYN>
        	

         
        	

         
        	

         
        	

         
        	

         
        	

         
        	

         
    min  number
        Set the value of MIN to number. MIN is used in non-canonical mode
input processing ( icanon).
    time  number
        Set the value of TIME to number. TIME is used in non-canonical mode
input processing ( icanon).

    Combination Modes

    saved settings
        Set the current terminal characteristics to the saved settings
produced by the -g option.
    evenp or parity
        Enable parenb and cs7; disable parodd.
    oddp
        Enable parenb, cs7, and parodd.
    -parity, -evenp, or -oddp
        Disable parenb, and set cs8.
    raw  (-raw or cooked)
        [XSI] [Option Start]
        Enable (disable) raw input and output. Raw mode shall be equivalent
to setting:

        stty cs8 erase ^- kill ^- intr ^- \
            quit ^- eof ^- eol ^- -post -inpck   

    nl  (-nl)
        Disable (enable) icrnl. In addition, -nl unsets inlcr and igncr.
    ek
        Reset ERASE and KILL characters back to system defaults.
    sane
        Reset all modes to some reasonable, unspecified, values.

.Sh STDIN

    Although no input is read from standard input, standard input shall be
used to get the current terminal I/O characteristics and to set new terminal
I/O characteristics.

.Sh INPUT FILES

    None.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of stty:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        This variable determines the locale for the interpretation of
sequences of bytes of text data as characters (for example, single-byte as
opposed to multi-byte characters in arguments) and which characters are in
the class print.
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    If operands are specified, no output shall be produced.

    If the -g option is specified, stty shall write to standard output the
current settings in a form that can be used as arguments to another instance
of stty on the same system.

    If the -a option is specified, all of the information as described in the
OPERANDS section shall be written to standard output. Unless otherwise
specified, this information shall be written as <space>-separated tokens in
an unspecified format, on one or more lines, with an unspecified number of
tokens per line. Additional information may be written.

    If no options or operands are specified, an unspecified subset of the
information written for the -a option shall be written.

    If speed information is written as part of the default output, or if the
-a option is specified and if the terminal input speed and output speed are
the same, the speed information shall be written as follows:

    "speed %d baud;", <speed>

    Otherwise, speeds shall be written as:

    "ispeed %d baud; ospeed %d baud;", <ispeed>, <ospeed>

    In locales other than the POSIX locale, the word baud may be changed to
something more appropriate in those locales.

    If control characters are written as part of the default output, or if
the -a option is specified, control characters shall be written as:

    "%s = %s;", <control-character name>, <value>

    where <value> is either the character, or some visual representation of
the character if it is non-printable, or the string undef if the character is
disabled.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        The terminal options were read or set successfully.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    The -g flag is designed to facilitate the saving and restoring of
terminal state from the shell level. For example, a program may:

    saveterm="$(stty -g)"       # save terminal state
    stty (new settings)         # set new state
    ...                         # ...

NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



stty.c

/*
 * Copyright 2008-2009 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <termios.h>
#include <libpu.h>

#define COMPACT_PFX "PUSTTY1:"

enum {
	stty_cfl		= 0,
	stty_ispeed		= 1,
	stty_ospeed		= 2,
	stty_ifl		= 3,
	stty_ofl		= 4,
	stty_lfl		= 5,
	stty_cchar		= 6,
};

static const struct stty_param {
	const char		*name;		/* name on cmd line */
	int			ptype;		/* type of param */
	tcflag_t		val;		/* value to set */
	tcflag_t		val_clear;	/* value to clear */
} params[] = {
	/*
	 * control flags
	 */
	{ "parenb", stty_cfl, PARENB, PARENB },
	{ "parodd", stty_cfl, PARODD, PARODD },

	{ "cs5", stty_cfl, CS5, CSIZE },
	{ "cs6", stty_cfl, CS6, CSIZE },
	{ "cs7", stty_cfl, CS7, CSIZE },
	{ "cs8", stty_cfl, CS8, CSIZE },

	{ "ispeed", stty_ispeed, },
	{ "ospeed", stty_ospeed, },

	{ "hupcl", stty_cfl, HUPCL, HUPCL },
	/* FIXME: support "hup", alias for "hupcl" */

	{ "cstopb", stty_cfl, CSTOPB, CSTOPB },
	{ "cread", stty_cfl, CREAD, CREAD },
	{ "clocal", stty_cfl, CLOCAL, CLOCAL },

	/*
	 * input flags
	 */
	{ "ignbrk", stty_ifl, IGNBRK, IGNBRK },
	{ "brkint", stty_ifl, BRKINT, BRKINT },
	{ "ignpar", stty_ifl, IGNPAR, IGNPAR },
	{ "parmrk", stty_ifl, PARMRK, PARMRK },
	{ "inpck", stty_ifl, INPCK, INPCK },
	{ "istrip", stty_ifl, ISTRIP, ISTRIP },
	{ "inlcr", stty_ifl, INLCR, INLCR },
	{ "igncr", stty_ifl, IGNCR, IGNCR },
	{ "icrnl", stty_ifl, ICRNL, ICRNL },
	{ "ixon", stty_ifl, IXON, IXON },
	{ "ixany", stty_ifl, IXANY, IXANY },
	{ "ixoff", stty_ifl, IXOFF, IXOFF },

	/*
	 * output flags
	 */
	{ "opost", stty_ofl, OPOST, OPOST },
	{ "ocrnl", stty_ofl, OCRNL, OCRNL },
	{ "onocr", stty_ofl, ONOCR, ONOCR },
	{ "onlret", stty_ofl, ONLRET, ONLRET },
	{ "ofill", stty_ofl, OFILL, OFILL },
	{ "ofdel", stty_ofl, OFDEL, OFDEL },

	{ "cr0", stty_ofl, CR0, CRDLY },
	{ "cr1", stty_ofl, CR1, CRDLY },
	{ "cr2", stty_ofl, CR2, CRDLY },
	{ "cr3", stty_ofl, CR3, CRDLY },

	{ "nl0", stty_ofl, NL0, NLDLY },
	{ "nl1", stty_ofl, NL1, NLDLY },

	/* FIXME: support "tabs", alias for "tab0" */
	{ "tab0", stty_ofl, TAB0, TABDLY },
	{ "tab1", stty_ofl, TAB1, TABDLY },
	{ "tab2", stty_ofl, TAB2, TABDLY },
	{ "tab3", stty_ofl, TAB3, TABDLY },

	{ "bs0", stty_ofl, BS0, BSDLY },
	{ "bs1", stty_ofl, BS1, BSDLY },

	{ "ff0", stty_ofl, FF0, FFDLY },
	{ "ff1", stty_ofl, FF1, FFDLY },

	{ "vt0", stty_ofl, VT0, VTDLY },
	{ "vt1", stty_ofl, VT1, VTDLY },

	/*
	 * local modes
	 */
	{ "isig", stty_lfl, ISIG, ISIG },
	{ "icanon", stty_lfl, ICANON, ICANON },
	{ "iexten", stty_lfl, IEXTEN, IEXTEN },
	{ "echo", stty_lfl, ECHO, ECHO },
	{ "echoe", stty_lfl, ECHOE, ECHOE },
	{ "echok", stty_lfl, ECHOK, ECHOK },
	{ "echonl", stty_lfl, ECHONL, ECHONL },
	{ "noflsh", stty_lfl, NOFLSH, NOFLSH },
	{ "tostop", stty_lfl, TOSTOP, TOSTOP },

	/*
	 * control characters
	 */
	{ "eof", stty_cchar, VEOF },
	{ "eol", stty_cchar, VEOL },
	{ "erase", stty_cchar, VERASE },
	{ "intr", stty_cchar, VINTR },
	{ "kill", stty_cchar, VKILL },
	{ "quit", stty_cchar, VQUIT },
	{ "susp", stty_cchar, VSUSP },
	{ "start", stty_cchar, VSTART },
	{ "stop", stty_cchar, VSTOP },

	/* FIXME: handle multi-param aliases...
		evenp
		parity
		oddp
		-parity, -evenp, -oddp
		raw, -raw, cooked
		nl, -nl
		ek,
		sane
	 */
};

static struct termios ti;

static int usage(void)
{
	fprintf(stderr,
"stty [-a | -g]\n"
"	or\n"
"stty operands...\n");
	return 1;
}

static int stty_push_ti(void)
{
	if (tcsetattr(STDIN_FILENO, TCSANOW, &ti)) {
		perror("stty(tcsetattr)");
		return 1;
	}

	return 0;
}

static int stty_show(void)
{
	return 1;	/* FIXME: implement stty settings display */
}

static int stty_show_compact(void)
{
	printf(COMPACT_PFX "%x:%x:%x:%x:%u:%u:%u:%u:%u:%u:%u:%u:%u:%u:%u\n",
	       ti.c_iflag,
	       ti.c_oflag,
	       ti.c_cflag,
	       ti.c_lflag,
	       ti.c_ispeed,
	       ti.c_ospeed,
	       ti.c_cc[VEOF],
	       ti.c_cc[VEOF],
	       ti.c_cc[VERASE],
	       ti.c_cc[VINTR],
	       ti.c_cc[VKILL],
	       ti.c_cc[VQUIT],
	       ti.c_cc[VSUSP],
	       ti.c_cc[VSTART],
	       ti.c_cc[VSTOP]);
	return 0;
}

static int stty_set_compact(const char *settings)
{
	int rc, c_cc[20], i;

	rc = sscanf(settings,
		    COMPACT_PFX "%x:%x:%x:%x:%u:%u:%u:%u:%u:%u:%u:%u:%u:%u:%u\n",
		    &ti.c_iflag,
		    &ti.c_oflag,
		    &ti.c_cflag,
		    &ti.c_lflag,
		    &ti.c_ispeed,
		    &ti.c_ospeed,
		    &c_cc[VEOF],
		    &c_cc[VEOF],
		    &c_cc[VERASE],
		    &c_cc[VINTR],
		    &c_cc[VKILL],
		    &c_cc[VQUIT],
		    &c_cc[VSUSP],
		    &c_cc[VSTART],
		    &c_cc[VSTOP]);
	if (rc != 15) {
		fprintf(stderr, _("stty: invalid compact settings format\n"));
		return 1;
	}

	for (i = 0; i <= 16; i++)
		ti.c_cc[i] = (unsigned char) c_cc[i];

	return stty_push_ti();
}

static const struct stty_param *last_param;

static int param_apply(const struct stty_param *param, const char *setting,
		       bool set_val)
{
	switch (param->ptype) {
	case stty_cfl:
		ti.c_cflag &= ~param->val_clear;
		if (set_val)
			ti.c_cflag |= param->val;
		break;
	case stty_ifl:
		ti.c_iflag &= ~param->val_clear;
		if (set_val)
			ti.c_iflag |= param->val;
		break;
	case stty_ofl:
		ti.c_oflag &= ~param->val_clear;
		if (set_val)
			ti.c_oflag |= param->val;
		break;
	case stty_lfl:
		ti.c_lflag &= ~param->val_clear;
		if (set_val)
			ti.c_lflag |= param->val;
		break;

	case stty_ispeed:
	case stty_ospeed:
	case stty_cchar:
		last_param = param;
		return -1;	/* obtain next arg */

	default:
		return 1;
	}

	return 0;
}

static int param_cchar(const struct stty_param *param, const char *setting)
{
	return 1;		/* FIXME: interpret control char setting */
}

#define X(speed) \
	case (speed): return B##speed;

static unsigned int speed_to_bits(unsigned int speed)
{
	switch (speed) {
	X(0)
	X(50)
	X(75)
	X(110)
	X(134)
	X(150)
	X(200)
	X(300)
	X(600)
	X(1200)
	X(1800)
	X(2400)
	X(4800)
	X(9600)
	X(19200)
	X(38400)
	X(57600)
	X(115200)
	X(230400)
	X(460800)
	X(500000)
	X(576000)
	X(921600)
	X(1000000)
	X(1152000)
	X(1500000)
	X(2000000)
	X(2500000)
	X(3000000)
	X(3500000)
	X(4000000)

	default:
		break;
	}

	return 0xffffffffU;
}

#undef X

static int param_speed(unsigned int speed, bool is_input)
{
	unsigned int bits = speed_to_bits(speed);
	int rc;

	if (bits == 0xffffffffU)
		return 1;

	if (is_input)
		rc = cfsetispeed(&ti, bits);
	else
		rc = cfsetospeed(&ti, bits);
	
	if (rc < 0)
		return 1;
	
	return 0;
}

static int stty_set(const char *setting)
{
	bool set_val = true;
	int i;
	unsigned int speed;

	if (!setting) {
		fprintf(stderr, _("stty: '%s' missing required argument\n"),
			last_param ? last_param->name : "(null)");
		return 1;
	}

	if (last_param) {
		const struct stty_param *param;
		unsigned int speed;

		param = last_param;
		last_param = NULL;

		switch (param->ptype) {
		case stty_ispeed:
			if (sscanf(setting, "%u", &speed) == 1)
				return param_speed(speed, true);
			break;
		case stty_ospeed:
			if (sscanf(setting, "%u", &speed) == 1)
				return param_speed(speed, false);
			break;
		case stty_cchar:
			return param_cchar(param, setting);
		}

		goto err_out;
	}

	if (*setting == '-') {
		set_val = false;
		setting++;
	}

	for (i = 0; i < ARRAY_SIZE(params); i++) {
		if (!strcmp(setting, params[i].name))
			return param_apply(&params[i], setting, set_val);
	}

	if (sscanf(setting, "%u", &speed) == 1) {
		if (param_speed(speed, true))
			return 1;
		if (param_speed(speed, false))
			return 1;
		return 0;
	}

err_out:
	fprintf(stderr, _("stty: invalid argument '%s'\n"), setting);
	return 1;
}

static int is_help_arg(const char *arg)
{
	if (!strcmp(arg, "-h") ||
	    !strcmp(arg, "--help") ||
	    !strcmp(arg, "-v") ||
	    !strcmp(arg, "-V") ||
	    !strcmp(arg, "-H") ||
	    !strcmp(arg, "-?"))
		return 1;

	return 0;
}

static bool is_compact_form(const char *arg)
{
	return (strncmp(arg, COMPACT_PFX, strlen(COMPACT_PFX)) == 0);
}

int main (int argc, char *argv[])
{
	int i, rc;

	pu_init();

	if ((argc == 2) && (is_help_arg(argv[1])))
		return usage();

	if (tcgetattr(STDIN_FILENO, &ti)) {
		perror("stty(tcgetattr)");
		return 1;
	}

	if (argc == 1)
		return stty_show();
	else if ((argc == 2) && (!strcmp(argv[1], "-a")))
		return stty_show();

	else if ((argc == 2) && (!strcmp(argv[1], "-g")))
		return stty_show_compact();
	else if ((argc == 2) && (is_compact_form(argv[1])))
		return stty_set_compact(argv[1]);

	for (i = 1; i < argc; i++) {
		rc = stty_set(argv[i]);
		if (rc > 0)
			return rc;
		if (rc < 0) {
			i++;
			rc = stty_set(i >= argc ? NULL : argv[i]);
			if (rc)
				return rc;
		}
	}

	return stty_push_ti();
}

